src/doc/src/reference/features.md

```diff
@@ -1,134 +1,456 @@
+  > the selected packages defines a given feature.
 
-* grouping a number of low-level optional dependencies together into a single
-  high-level feature;
-* specifying packages that are recommended (or suggested) to be included by
-  users of the package; and
-* including a feature (like `secure-password` in the motivating example) that
-  will only work if an optional dependency is available, and would be difficult
-  to implement as a separate package (for example, it may be overly difficult to
-  design an IO package to be completely decoupled from OpenSSL, with opt-in via
-  the inclusion of a separate package).
+* `--all-features`: Activates all features of all packages selected on the
+  command-line.
 
-In almost all cases, it is an antipattern to use these features outside of
-high-level packages that are designed for curation. If a feature is optional, it
-can almost certainly be expressed as a separate package.
+* `--no-default-features`: Does not activate the [`default`
+  feature](#the-default-feature) of the selected packages.
 
-[crates.io]: https://crates.io/
+  > **Note**: With the `resolver = "1"` option, this only applies to the
+  > package in the current directory.
+
+[workspace]: workspaces.md
+
+### Feature unification
+
+Features are unique to the package that defines them. Enabling a feature on a
+package does not enable a feature of the same name on other packages.
+
+When a dependency is used by multiple packages, Cargo will use the union of
+all features enabled on that dependency when building it. This helps ensure
+that only a single copy of the dependency is used. See the [features section]
+of the resolver documentation for more details.
+
+For example, let's look at the [`winapi`] package which uses a [large
+number][winapi-features] of features. If your package depends on a package
+`foo` which enables the "fileapi" and "handleapi" features of `winapi`, and
+another dependency `bar` which enables the "std" and "winnt" features of
+`winapi`, then `winapi` will be built with all four of those features enabled.
+
+![winapi features example](../images/winapi-features.svg)
+
+[`winapi`]: https://crates.io/crates/winapi
+[winapi-features]: https://github.com/retep998/winapi-rs/blob/0.3.9/Cargo.toml#L25-L431
+
+> **Note**: Sometimes feature unification may be a little too aggressive. See
+> the `resolver = "2"` option [described below](#feature-resolver-version-2)
+> for improved behavior.
+
+A consequence of this is that features should be *additive*. That is, enabling
+a feature should not disable functionality, and it should usually be safe to
+enable any combination of features. A feature should not introduce a
+[SemVer-incompatible change](#semver-compatibility).
+
+For example, if you want to optionally support [`no_std`] environments, **do
+not** use a `no_std` feature. Instead, use a `std` feature that *enables*
+`std`. For example:
+
+```rust
+#![cfg_attr(not(feature = "std"), no_std)]
```

Comment:
In the spirit of additiveness, I think it'd be better to show here the unconditional `#![no_std]` followed by a conditional `use std::...` perhaps?